### ISSUE TITLE
Support for function decorators

### ISSUE THREAD

**Issue by khia** [elixir-lang/elixir#448]
Some languages has features to extract code snippets from documentations of a module in order to run them. This way you can write your examples of usage of the function without a need to write a test case for it.

For basic information you could read http://docs.python.org/library/doctest.html

There is an experimental implementation of doctests for elixir here https://github.com/khia/elixir-doctest
It is based on before_compile hooks however this approach has one significant limitation. Since in this case we run tests in different scope than a module we are testing we have to prefix functions with module name and we cannot call private functions.

**Comment by orestis** [comment]
Hi, sorry for revisiting a 4 year old issue, but if the point was to reproduce parts of Python's decorators, I cannot understand how the on_definition callback achieves this - it has no power to change the function being defined.

To be able to reproduce Python's behavior, the callback should *return* something that would replace the function body. If this is not the aim of the on_definition callback, it's ok, it should just be documented, even in this issue which is quite high in the Google results of "elixir decorators".

**Comment by josevalim** [comment]
@orestis there are a couple differences in Elixir that makes it harder to implement exactly like python decorators. For example, if you add `@foo bar` before a function, do you mean to apply the decorator that particular clause or all clauses for that function?

It is most likely that you want the latter and therefore rewriting right on the `@on_definition` callback would not be the way to go. Instead, it would be better to store the information in a module attribute and use a `@before_compile` callback to do any changes. It is even possible to replace the function body by using `defoverridable`.

While it can certainly be done, it sounds complicated (and I believe that's a good thing). We tend to prefer function composition and still Elixir already provides a powerful macro system (which is at least lexical). So not having yet another mechanism for indirection is a good thing.

TL;DR - Elixir provides different solutions to where one would use Python decorators. If you still need them, it can be pulled off with a combination of `@on_definition`, `@before_compile` and module attributes.
